pkg/runner/runner.go

```diff
@@ -0,0 +1,346 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package runner
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/run/tracer"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/run/types"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/oci"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+type runnerState string
+
+const (
+	RunnerStateNotStarted runnerState = "not-started"
+	RunnerStateStarted    runnerState = "started"
+	RunnerStateCompleted  runnerState = "completed"
+)
+
+type RunnerOpts struct {
+	tracer.RunOpts
+
+	Timeout    time.Duration
+	context    context.Context
+	ebpfParams map[string]types.EBPFParam
+}
+
+// WithTimeout sets the timeout for the gadget
+func WithTimeout(timeout time.Duration) func(*RunnerOpts) {
+	return func(r *RunnerOpts) {
+		r.Timeout = timeout
+	}
+}
+
+// WithContext sets the context for the gadget
+// It is useful if one wants to cancel the gadget execution
+func WithContext(ctx context.Context) func(*RunnerOpts) {
+	return func(r *RunnerOpts) {
+		r.context = ctx
+	}
+}
+
+// WithValidateMetadata sets whether the metadata should be validated against the eBPF program
+func WithValidateMetadata(validateMetadata bool) func(*RunnerOpts) {
+	return func(r *RunnerOpts) {
+		r.ValidateMetadata = validateMetadata
+	}
+}
+
+// WithPullPolicy sets the pull policy for pulling the gadget image
+func WithPullPolicy(pullPolicy string) func(*RunnerOpts) {
+	return func(r *RunnerOpts) {
+		r.PullPolicy = pullPolicy
+	}
+}
+
+// WithAuthOpts sets the authentication options when pulling the gadget image
+func WithAuthOpts(authOpts *oci.AuthOptions) func(*RunnerOpts) {
+	return func(r *RunnerOpts) {
+		r.AuthOpts = authOpts
+	}
+}
+
+type Runner struct {
+	eventBuffer
+
+	RunnerOpts RunnerOpts
+	tracer     tracerInterface
+
+	state      runnerState
+	stateMutex sync.Mutex
+	stateCV    *sync.Cond
+}
+
+// eventBuffer implements the logger.Logger interface
+type eventBuffer struct {
+	// TODO: buffersize/maxsize?
+	buffer          []string
+	eventPoisonPill string
+	eventMutex      sync.Mutex
+	eventCV         *sync.Cond
+}
+
+func (r *eventBuffer) Output(payload string) {
+	r.eventMutex.Lock()
+	defer r.eventMutex.Unlock()
+	r.buffer = append(r.buffer, payload)
+	r.eventCV.Broadcast()
+}
+
+func (r *eventBuffer) Logf(severity logger.Level, fmtStr string, params ...any) {
+	fmt.Printf(fmtStr, params...)
+}
+
+type tracerInterface interface {
+	InitWithGadgetInfo(*types.GadgetInfo) error
+	RunWithoutGadgetContext(logger logger.Logger, ctx context.Context, timeout time.Duration, params *params.Params) error
+	Close()
+}
+
+// NewRunner creates a new runner with the given image and options
+func NewRunner(image string, opts ...func(*RunnerOpts)) (*Runner, error) {
```

Comment:
Functions like `NewRunner()` could have examples in code blocks too:
https://tip.golang.org/doc/comment#code